planner/core/logical_plan_builder.go

```diff
@@ -4226,6 +4226,21 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 		}
 	}
 
+	// check whether the path's index has a tidb_shard() prefix and the index column count more than 1
+	// e.g. index(tidb_shard(a), a)
+	for _, path := range ds.possibleAccessPaths {
+		if !path.IsTablePath() {
+			col := expression.IndexCol2Col(ds.Columns, ds.schema.Columns, path.Index.Columns[0])
+			if col != nil &&
+				expression.GcColumnExprIsTidbShard(col.VirtualExpr) &&
```

Comment:
Could we add a flag indicating if it's a shard index so we don't need to much check here, also it's easier to control the actual meaning of `shared index` type.

Reply:
I agree with you fully. At first，I want to add a new syntax `SHARD INDEX` and add a flag for table meta so that it's easier to judge whether it has a shard index and whether the index is shard index. But as you know , we every body consider that it can cause unnecessary trouble for the fully realized shard index by @tiancaiamao , I give it up at last. If I add a flag by expression `tidb_shard` not by key word `SHARD`, I will get much information about this expression and do the check for it as now at ddll phase, I think it's not suitable . e.g. primary key is a meta flag for a index by key word 'PRIMARY'.